File: problems/1110-delete-nodes-and-return-forest/post-order.cpp
/**
 * 后序遍历
 * 
 * 时间：`O(N)`, 16ms
 */

class Solution {
public:
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& toDeletes) {
        delSet.insert(toDeletes.begin(), toDeletes.end()); // 哈希
        root = dfs(root);
        if (root) res.push_back(root);
        return res;
    }

    TreeNode* dfs(TreeNode* node) {
        if (!node) return node;

        node->left = dfs(node->left);
        node->right = dfs(node->right);
        
        if (delSet.count(node->val)) { // 本结点需要被删除
            if (node->left) res.push_back(node->left); // 分割左子树
            if (node->right) res.push_back(node->right); // 分割右子树
            return NULL; // 删除本结点
        } else {
            return node;
        }
    }

private:
    vector<TreeNode*> res;
    unordered_set<int> delSet;
};

File: biweekly-contest/biweekly-contest-21/3-longest-zigzag-path-in-a-binary-tree/post-order.js
/**
 * 后序遍历（DP）
 */
var longestZigZag = function(root) {
  let res = 0;

  // 返回数组，表示该结点往左、右走的最长路径
  function helper(node) {
    if (!node) return [-1, -1];

    const [ll, lr] = helper(node.left);
    const [rl, rr] = helper(node.right);
    res = Math.max(res, 1 + lr, 1 + rl); // 记录答案
    return [1 + lr, 1 + rl];
  }

  helper(root);
  return res;
};